File: problem/Min_Cost_Path/minCost.c
#include <stdlib.h>
#include <string.h>
#include <limits.h>

typedef unsigned long long u64;

typedef struct Edge
{
    int to;
    int weight;
    struct Edge *next;
} Edge;

typedef struct
{
    u64 *data;
    int size;
    int capacity;
} MinHeap;

MinHeap *createHeap(int cap)
{
    MinHeap *h = (MinHeap *)malloc(sizeof(MinHeap));
    h->data = (u64 *)malloc(sizeof(u64) * (cap + 1));
    h->size = 0;
    h->capacity = cap;
    return h;
}

void push(MinHeap *h, u64 val)
{
    int i = ++h->size;
    while (i > 1 && h->data[i >> 1] > val)
    {
        h->data[i] = h->data[i >> 1];
        i >>= 1;
    }
    h->data[i] = val;
}

u64 pop(MinHeap *h)
{
    u64 ret = h->data[1];
    u64 last = h->data[h->size--];
    int i = 1, child = 2;
    while (child <= h->size)
    {
        if (child < h->size && h->data[child + 1] < h->data[child])
            child++;
        if (last <= h->data[child])
            break;
        h->data[i] = h->data[child];
        i = child;
        child <<= 1;
    }
    h->data[i] = last;
    return ret;
}

int minCost(int n, int **edges, int edgesSize, int *edgesColSize)
{
    Edge **adj = (Edge **)calloc(n, sizeof(Edge *));
    for (int i = 0; i < edgesSize; i++)
    {
        int u = edges[i][0], v = edges[i][1], w = edges[i][2];
        Edge *e1 = (Edge *)malloc(sizeof(Edge));
        e1->to = v;
        e1->weight = w;
        e1->next = adj[u];
        adj[u] = e1;
        Edge *e2 = (Edge *)malloc(sizeof(Edge));
        e2->to = u;
        e2->weight = w * 2;
        e2->next = adj[v];
        adj[v] = e2;
    }

    unsigned *dist = (unsigned *)malloc(sizeof(unsigned) * n);
    memset(dist, 255, sizeof(unsigned) * n);

    MinHeap *pq = createHeap(edgesSize * 2 + n);

    dist[0] = 0;
    push(pq, 0);

    int res = -1;

    while (pq->size > 0)
    {
        u64 curr = pop(pq);
        unsigned d = curr >> 16;
        int u = curr & 0xFFFF;

        if (d > dist[u])
            continue;
        if (u == n - 1)
        {
            res = d;
            break;
        }

        for (Edge *e = adj[u]; e; e = e->next)
        {
            unsigned d2 = d + e->weight;
            if (d2 < dist[e->to])
            {
                dist[e->to] = d2;
                push(pq, ((u64)d2 << 16) | e->to);
            }
        }
    }

    return res;
}

File: problem/Min_Cost_Path/minCost.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var minCost = function(n, edges) {
    const adj = Array.from({ length: n }, () => []);
    for (const [u, v, w] of edges) {
        adj[u].push([w, v]);
        adj[v].push([2 * w, u]);
    }

    const minHeap = [];
    
    function push(d, u) {
        minHeap.push([d, u]);
        let i = minHeap.length - 1;
        while (i > 0) {
            const p = (i - 1) >>> 1;
            if (minHeap[p][0] <= minHeap[i][0]) break;
            [minHeap[p], minHeap[i]] = [minHeap[i], minHeap[p]];
            i = p;
        }
    }

    function pop() {
        if (minHeap.length === 0) return null;
        const res = minHeap[0];
        const last = minHeap.pop();
        if (minHeap.length > 0) {
            minHeap[0] = last;
            let i = 0;
            const len = minHeap.length;
            while (true) {
                let left = (i << 1) + 1;
                let right = left + 1;
                let swap = null;

                if (left < len) {
                    if (minHeap[left][0] < minHeap[i][0]) swap = left;
                }
                if (right < len) {
                    if ((swap === null && minHeap[right][0] < minHeap[i][0]) ||
                        (swap !== null && minHeap[right][0] < minHeap[swap][0])) {
                        swap = right;
                    }
                }
                if (swap === null) break;
                [minHeap[i], minHeap[swap]] = [minHeap[swap], minHeap[i]];
                i = swap;
            }
        }
        return res;
    }

    const dist = new Uint32Array(n).fill(-1);
    dist[0] = 0;
    push(0, 0);

    while (minHeap.length > 0) {
        const [d, u] = pop();

        if (d > dist[u]) continue;
        if (u === n - 1) return d;

        for (const [w, v] of adj[u]) {
            const d2 = d + w;
            if (d2 < dist[v]) {
                dist[v] = d2;
                push(d2, v);
            }
        }
    }

    return -1;
};